lib/core/unittest/CStaticThreadPoolTest.cc

```diff
@@ -202,7 +204,46 @@ BOOST_AUTO_TEST_CASE(testWithExceptions) {
     }
 
     // We didn't lose any real tasks.
-    BOOST_REQUIRE_EQUAL(200u, counter.load());
+    BOOST_REQUIRE_EQUAL(200, counter.load());
+}
+
+BOOST_AUTO_TEST_CASE(testNumberThreadsInUse) {
+
+    // Start a threadpool then change the number of threads and check we aren't
+    // getting execution on more than the specified number of distinct threads.
+
+    using TThreadIdUSet = boost::unordered_set<std::thread::id, std::hash<std::thread::id>>;
+
+    core::CStaticThreadPool pool{8};
+
+    std::mutex mutex;
+    std::size_t numberProcessedTasks{0};
+    TThreadIdUSet executionThreads;
+
+    for (std::size_t numberThreadsInUse : {5, 6, 2}) {
+
+        pool.numberThreadsInUse(numberThreadsInUse);
+
+        for (std::size_t i = 0; i < 200; ++i) {
+            pool.schedule([&] {
+                std::scoped_lock<std::mutex> lock{mutex};
+                ++numberProcessedTasks;
+                executionThreads.insert(std::this_thread::get_id());
+            });
+        }
+
+        for (;;) {
+            std::scoped_lock<std::mutex> lock{mutex};
+            if (numberProcessedTasks == 200) {
+                LOG_DEBUG(<< "# threads used = " << executionThreads.size());
+                // A subset of threads can steal all the work.
+                BOOST_REQUIRE(numberThreadsInUse <= executionThreads.size());
```

Comment:
Shouldn't we be asserting that `executionThreads.size() <= numberThreadsInUse` instead?

Reply:
We should indeed!